### PR TITLE
Storage notifications routed via the constellation.

### LINKED ISSUE #5196: Dispatch StorageEvent events to other windows when modifying storage instances
Spec: https://html.spec.whatwg.org/multipage/webstorage.html#the-storageevent-interface, https://html.spec.whatwg.org/multipage/webstorage.html#send-a-storage-notification, https://html.spec.whatwg.org/multipage/webstorage.html#the-sessionstorage-attribute:send-a-storage-notification, https://html.spec.whatwg.org/multipage/webstorage.html#the-localstorage-attribute:send-a-storage-notification
Code: http://mxr.mozilla.org/servo/source/components/script/dom/storage.rs

The bit about "for every Document object whose Window object's localStorage attribute's Storage object is associated with the same storage area, other than x, send a storage notification" is the important bit to figure out here.

### PR DESCRIPTION
**Pull request by asajeffrey**
At the moment, storage notifications are only sent within a script thread, not to same-origin pipelines in different script threads. This PR fixes that.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #5196
- [X] These changes do not require tests because existing tests now pass

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14023)

### CODE REVIEW

File: components/constellation/constellation.rs
Comment (nox) [inline components/constellation/constellation.rs]: Why ignore failure here?

File: components/constellation/constellation.rs
Comment (nox) [inline components/constellation/constellation.rs]: Nit: remove parens.

File: components/constellation/constellation.rs
Comment (nox) [inline components/constellation/constellation.rs]: <strike>Why ignore failure here? Why isn't that method taking an `Url` directly? I see below that the `String` comes directly from an `Url`, so please just unwrap the result of `Url::parse` or pass an actual `Url` around.</strike>

Oh I see you did that in second commit. Could you squash them?

File: components/script/dom/storage.rs
Comment (nox) [inline components/script/dom/storage.rs]: Why the sudden `*`?

File: components/script_traits/script_msg.rs
Comment (nox) [inline components/script_traits/script_msg.rs]: Maybe use a struct variant instead of documenting each field in prose?

**Review by nox — CHANGES_REQUESTED** [review]

**Comment by nox** [comment]
GitHub somehow decided that some of my comments are outdated when they are not, please expand them when you read this @asajeffrey.

File: components/constellation/constellation.rs
Comment (asajeffrey) [inline components/constellation/constellation.rs]: I'll squash them before merging.

File: components/constellation/constellation.rs
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Hmm, I prefer explicit parens rather than relying on operator pecedence to resolve ambiguity.

File: components/constellation/constellation.rs
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Fixed.

**Comment by asajeffrey** [comment]
We should probably hold off on merging this until #14013 has landed, c.f. https://github.com/servo/servo/pull/14013/files#diff-ab86862fdf6e275f1f412558ce78bc2aR195.

File: components/script_traits/script_msg.rs
Comment (asajeffrey) [inline components/script_traits/script_msg.rs]: I agree with you, but we should be consistent about it, not just adopt struct variants piecemeal.

File: components/script/dom/storage.rs
Comment (asajeffrey) [inline components/script/dom/storage.rs]: Apparently no good reason, not sure how this snuck in. Fixed.

**Comment by nox** [comment]
Squash and r=me.

**Comment by asajeffrey** [comment]
Squashed. @bors-servo r=nox

**Pull request merged**
